### PR TITLE
[Self-Driving] Index memory consumption

### PR DESCRIPTION
**Pull request by adi64**
In order to re-integrate the results of the self-driving PR #715 together with the new plugin architecture, here's the first round of changes: Indexes can now report their memory consumption and can estimate memory consumption for a hypothetical index.

This straightforward implementation modifies the index structures themselves and is not part of a plugin.

Not all indexes are completely supported (no support means runtime errors):

- Group Key Index: Supported
- Composite Group Key Index: Supported
- B-Tree Index: Reporting supported, estimation not supported
- Adaptive Radix Tree Index: Neither reporting nor estimation supported

### CODE REVIEW

File: src/lib/storage/index/base_index.hpp [hunk 1]
```diff
@@ -38,6 +38,19 @@ class BaseIndex : private Noncopyable {
   // For now we use an iterator over a vector of chunkoffsets as the GroupKeyIndex works like this
   using Iterator = std::vector<ChunkOffset>::const_iterator;
 
+  /**
+   * Predicts the memory consumption in bytes of creating an index with the specific index implementation <type>
+   * on a Chunk with the following statistics:
+   *
+   * row_count - overall number of rows
+   * value_count - number of distinct values
```
Comment (Bensk1) [inline src/lib/storage/index/base_index.hpp]: For example, in the histograms we use `distinct_count` and probably at other places throughout the code. Should we stick with that?

File: src/lib/storage/index/base_index.hpp [hunk 2]
```diff
@@ -38,6 +38,19 @@ class BaseIndex : private Noncopyable {
...
+   * value_bytes - (average) size of a single value in bytes
+   *
+   * If no prediction is possible, this shall fail.
```
Comment (Bensk1) [inline src/lib/storage/index/base_index.hpp]: Is this intended for the other not yet supported index types? There, a prediction would be possible, but is simply not implemented. In all other cases, a prediction should be possible, right?

File: src/lib/storage/index/group_key/composite_group_key_index.cpp [hunk 3]
```diff
@@ -175,4 +180,11 @@ std::vector<std::shared_ptr<const BaseSegment>> CompositeGroupKeyIndex::_get_ind
   return result;
 }
 
+size_t CompositeGroupKeyIndex::_memory_consumption() const {
+  uintptr_t byte_count = _keys.size() * _keys.key_size();
```
Comment (Bensk1) [inline src/lib/storage/index/group_key/composite_group_key_index.cpp]: `uintptr_t`?

File: src/lib/storage/index/group_key/group_key_index.cpp [hunk 4]
```diff
@@ -91,4 +96,11 @@ std::vector<std::shared_ptr<const BaseSegment>> GroupKeyIndex::_get_indexed_segm
   return {_indexed_segments};
 }
 
+size_t GroupKeyIndex::_memory_consumption() const {
+  uintptr_t bytes = sizeof(_indexed_segments);
```
Comment (Bensk1) [inline src/lib/storage/index/group_key/group_key_index.cpp]: s.o.

File: src/lib/storage/index/base_index.hpp (hunk 1 above)
Comment (Bouncner) [reply src/lib/storage/index/base_index.hpp]: For me, `value_count` is not the same as a `distinct_count`.

Comment (adi64) [reply src/lib/storage/index/base_index.hpp]: changed to `distinct_count`.

File: src/lib/storage/index/base_index.hpp (hunk 2 above)
Comment (adi64) [reply src/lib/storage/index/base_index.hpp]: Yes. I changed the comment to "If no prediction is possible (or it is not implemented yet), this shall fail" to reflect that.

File: src/lib/storage/index/group_key/composite_group_key_index.cpp (hunk 3 above)
Comment (adi64) [reply src/lib/storage/index/group_key/composite_group_key_index.cpp]: changed to `size_t`

File: src/lib/storage/index/adaptive_radix_tree/adaptive_radix_tree_index.cpp [hunk 5]
```diff
@@ -16,6 +16,12 @@
 
 namespace opossum {
 
+size_t AdaptiveRadixTreeIndex::estimate_memory_consumption(ChunkOffset row_count, ChunkOffset distinct_count,
+                                                           uint32_t value_bytes) {
+  // ToDo(anyone): If you use this index in combination with the Tuning subsystem, you need to properly implement this.
```
Comment (mrks) [inline src/lib/storage/index/adaptive_radix_tree/adaptive_radix_tree_index.cpp]: What is this todo good for? The fact that it's not implemented is obvious, the explanation of when you might need it feels redundant.

File: src/lib/storage/index/b_tree/b_tree_index.cpp [hunk 6]
```diff
@@ -5,14 +5,20 @@
...
+
 BTreeIndex::BTreeIndex(const std::vector<std::shared_ptr<const BaseSegment>>& segments_to_index)
     : BaseIndex{get_index_type_of<BTreeIndex>()}, _indexed_segments(segments_to_index[0]) {
   Assert((segments_to_index.size() == 1), "BTreeIndex only works with a single segment.");
   _impl =
       make_shared_by_data_type<BaseBTreeIndexImpl, BTreeIndexImpl>(_indexed_segments->data_type(), _indexed_segments);
 }
 
-uint64_t BTreeIndex::memory_consumption() const { return _impl->memory_consumption(); }
+size_t BTreeIndex::_memory_consumption() const { return _impl->memory_consumption(); }
```
Comment (mrks) [inline src/lib/storage/index/b_tree/b_tree_index.cpp]: Not part of this PR, but doesn't `BTreeIndexImpl<DataType>::memory_consumption()` ignore the heap size of `std::string`s?

Comment (adi64) [reply src/lib/storage/index/b_tree/b_tree_index.cpp]: I haven't looked into cpp-btree's implementation, so don't know what `_btree.bytes_used()` is actually doing. Could be that it just takes `sizeof(DataType)`. hm =/

**Comment by Bouncner** [comment]
Full CI?

**Comment by adi64** [comment]
Should I remove the Full CI label once I only push corrections from review?

**Comment by Bensk1** [comment]
No, you don't have to. The last commit before merging is required to build with FullCI activated. 

If you know that you are going to push 10 small not really relevant commits in a row, you could temporarily turn it off.

File: src/lib/storage/index/b_tree/b_tree_index.cpp (hunk 6 above)
Comment (mrks) [reply src/lib/storage/index/b_tree/b_tree_index.cpp]: Looks like it:

```
if (stats.leaf_nodes == 1 && stats.internal_nodes == 0) {
      return sizeof(*this) +
          sizeof(base_fields) + root()->max_count() * sizeof(value_type);
    }
```

For now, it's probably easiest to track the heap cost of the stored keys/values manually during construction of the index. I think we should get this in before we merge this PR, as we would otherwise report blatantly false values.

File: src/lib/storage/index/b_tree/b_tree_index_impl.cpp [hunk 7]
```diff
@@ -81,14 +83,30 @@ void BTreeIndexImpl<DataType>::_bulk_insert(const std::shared_ptr<const BaseSegm
...
 
+template <typename DataType>
+void BTreeIndexImpl<DataType>::_add_to_heap_memory_usage(const DataType&) {
+  // Except for std::string (see below), no supported data type uses heap allocations
+}
+
+template <>
+void BTreeIndexImpl<std::string>::_add_to_heap_memory_usage(const std::string& value) {
+  // Track only strings that are longer than the reserved stack space for short string optimization (SSO)
+  static const auto short_string_threshold = std::string("").capacity();
```
Comment (mrks) [inline src/lib/storage/index/b_tree/b_tree_index_impl.cpp]: nice
